packages/contracts/contracts/discovery/Staking.sol

```diff
@@ -73,77 +76,97 @@ contract Staking is IStaking {
     /// @inheritdoc IStaking
     function unbond(uint256 amount) external override(IStaking) {
         Stake storage _stake = stakes[msg.sender];
-        require(_stake.availableAmount >= amount, "not enough funds");
+        UnbondingList storage _unbondings = unbondings[msg.sender];
+
+        require(_stake.bonded >= amount, "not enough funds");
+
         Unbonding memory unbonding = Unbonding(
             amount,
             block.timestamp + UNBONDING_PERIOD
         );
 
-        _stake.unbondings[_stake.nextUnbonding] = unbonding;
-        _stake.nextUnbonding += 1;
-        _stake.availableAmount -= amount;
+        _unbondings.list[_unbondings.next] = unbonding;
+        _unbondings.next++;
+        _stake.bonded -= amount;
 
         emit Unbond(msg.sender, amount);
     }
 
     /// @inheritdoc IStaking
     function rebond(uint256 amount) external override(IStaking) {
         Stake storage _stake = stakes[msg.sender];
+        UnbondingList storage _unbondings = unbondings[msg.sender];
+
         require(
-            (_stake.actualAmount - _stake.availableAmount) >= amount,
+            (_stake.total - _stake.bonded) >= amount,
             "not enough unbonding funds"
         );
-        uint256 toBeRebonded = amount;
 
-        for (
-            uint256 i = _stake.nextUnbonding - 1;
-            i >= _stake.firstUnbonding;
-            i--
-        ) {
-            Unbonding storage unbonding = _stake.unbondings[i];
+        uint256 toBeRebonded = amount;
+        uint256 limit = _unbondings.first;
+        for (uint256 i = _unbondings.next - 1; i >= limit; i--) {
+            Unbonding storage unbonding = _unbondings.list[i];
 
             if (unbonding.amount >= toBeRebonded) {
                 unbonding.amount -= toBeRebonded;
-                _stake.availableAmount += toBeRebonded;
                 break;
             } else {
-                _stake.availableAmount += unbonding.amount;
                 toBeRebonded -= unbonding.amount;
-                delete _stake.unbondings[i];
-                _stake.nextUnbonding -= 1;
+                delete _unbondings.list[i];
+                _unbondings.next--;
             }
         }
+        _stake.bonded += amount;
 
         emit Rebond(msg.sender, amount);
     }
 
+    /// TODO add a gas meter to prevent this function from crashing if too many unbondings exist
     /// @inheritdoc IStaking
-    function withdraw(uint256 amount) external override(IStaking) {
+    function withdraw() external override(IStaking) {
```

Comment:
I don't see a use case for that. it's true you can decide to rebond 50% and withdraw the other 50%. but that's just as easily done with withdrawing 100% followed by a fresh stake

the only detail missing (which I left as a TODO note) is that we may need to add some check in case the array gets too big. although that's not a very realistic thing to happen